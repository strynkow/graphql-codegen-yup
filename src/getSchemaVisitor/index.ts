import { DeclarationBlock, indent } from '@graphql-codegen/visitor-plugin-common';
import {
  EnumTypeDefinitionNode,
  GraphQLSchema,
  InputObjectTypeDefinitionNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  UnionTypeDefinitionNode,
} from 'graphql';

import { TsVisitor } from '@graphql-codegen/typescript';
import { ValidationSchemaPluginConfig } from '../types';
import getFieldSchema from './getFieldSchema';

export default function getSchemaVisitor(schema: GraphQLSchema, config: ValidationSchemaPluginConfig) {
  const tsVisitor = new TsVisitor(schema, config);

  const importTypes: string[] = [];

  return {
    InputObjectTypeDefinition: (node: InputObjectTypeDefinitionNode) => {
      const name = tsVisitor.convertName(node.name.value);
      importTypes.push(name);

      const shape = node.fields
        ?.map((field) => getFieldSchema({ config, tsVisitor, schema, field, indentCount: 2 }))
        .join(',\n');

      return new DeclarationBlock({})
        .export()
        .asKind('function')
        .withName(`${name}Schema(): yup.SchemaOf<${name}>`)
        .withBlock([indent(`return yup.object({`), shape, indent('})')].join('\n')).string;
    },
    ObjectTypeDefinition: (node: ObjectTypeDefinitionNode) => {
      const name = tsVisitor.convertName(node.name.value);
      importTypes.push(name);

      const shape = node.fields
        ?.map((field) => getFieldSchema({ config, tsVisitor, schema, field, indentCount: 2 }))
        .join(',\n');

      return new DeclarationBlock({})
        .export()
        .asKind('function')
        .withName(`${name}Schema(): yup.SchemaOf<${name}>`)
        .withBlock(
          [
            indent(`return yup.object({`),
            indent(`__typename: yup.mixed().oneOf(['${node.name.value}', undefined]),`, 2),
            shape,
            indent('})'),
          ].join('\n')
        ).string;
    },
    EnumTypeDefinition: (node: EnumTypeDefinitionNode) => {
      const enumname = tsVisitor.convertName(node.name.value);
      importTypes.push(enumname);

      if (config.enumsAsTypes) {
        return new DeclarationBlock({})
          .export()
          .asKind('const')
          .withName(`${enumname}Schema`)
          .withContent(
            `yup.mixed().oneOf([${node.values?.map((enumOption) => `'${enumOption.name.value}'`).join(', ')}])`
          ).string;
      }

      const values = node.values
        ?.map(
          (enumOption) =>
            `${enumname}.${tsVisitor.convertName(enumOption.name, {
              useTypesPrefix: false,
              transformUnderscore: true,
            })}`
        )
        .join(', ');
      return new DeclarationBlock({})
        .export()
        .asKind('const')
        .withName(`${enumname}Schema`)
        .withContent(`yup.mixed().oneOf([${values}])`).string;
    },
    InterfaceTypeDefinition: (node: InterfaceTypeDefinitionNode) => {
      const name = tsVisitor.convertName(node.name.value);
      importTypes.push(name);

      const shape = node.fields
        ?.map((field) => getFieldSchema({ config, tsVisitor, schema, field, indentCount: 2 }))
        .join(',\n');

      return new DeclarationBlock({})
        .export()
        .asKind('function')
        .withName(`${name}Schema(): yup.SchemaOf<${name}>`)
        .withBlock(
          [
            indent(`return yup.object({`),
            indent(`__typename: yup.mixed().oneOf(['${node.name.value}', undefined]),`, 2),
            shape,
            indent('})'),
          ].join('\n')
        ).string;
    },
    UnionTypeDefinition: (node: UnionTypeDefinitionNode) => {
      const name = tsVisitor.convertName(node.name.value);
      importTypes.push(name);

      const shape = [];
      node.types.forEach((type) => {
        shape.push(indent(`case '${type?.name?.value}':`, 3));
        shape.push(indent(`return ${type?.name?.value}Schema();`, 4));
      });

      return new DeclarationBlock({})
        .export()
        .asKind('function')
        .withName(`${name}Schema(): any`)
        .withBlock(
          [
            indent(`return yup.lazy((value) => {`),
            indent('switch (value?.__typename) {', 2),
            ...shape,
            indent(`default:`, 3),
            indent(`throw new Error('Type not found.');`, 4),
            indent('}', 3),
            indent('})'),
          ].join('\n')
        ).string;
    },
    // Skip. These refer to top level op types, like mutation and query.
    // OperationTypeDefinition: (node: InterfaceTypeDefinitionNode) => {},
  };
}
