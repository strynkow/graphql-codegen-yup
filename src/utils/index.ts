import { ListTypeNode, NamedTypeNode, NonNullTypeNode, TypeNode } from 'graphql';

export function isListType(typ?: TypeNode): typ is ListTypeNode {
  return typ?.kind === 'ListType';
}

export function isNonNullType(typ?: TypeNode): typ is NonNullTypeNode {
  return typ?.kind === 'NonNullType';
}

export function isNamedType(typ?: TypeNode): typ is NamedTypeNode {
  return typ?.kind === 'NamedType';
}

export function isInput(kind: string): boolean {
  return kind.includes('Input');
}

export function maybeLazy(type: TypeNode, schema: string): string {
  if (isNamedType(type) && isInput(type.name.value)) {
    // TODO (orange): We have the tech to change this, after getting problems solved.
    return `yup.lazy(() => ${schema}) as never`;
  }
  return schema;
}
