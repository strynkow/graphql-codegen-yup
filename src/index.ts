import { PluginFunction, Types, oldVisit } from '@graphql-codegen/plugin-helpers';

import { GraphQLSchema } from 'graphql';
import { ValidationSchemaPluginConfig } from './types';
import getSchemaVisitor from './getSchemaVisitor';
import { transformSchemaAST } from '@graphql-codegen/schema-ast';

export const plugin: PluginFunction<ValidationSchemaPluginConfig, Types.ComplexPluginOutput> = (
  schema: GraphQLSchema,
  _documents: Types.DocumentFile[],
  config: ValidationSchemaPluginConfig
): Types.ComplexPluginOutput => {
  const { schema: _schema, ast } = transformSchemaAST(schema, config);
  const visitor = getSchemaVisitor(_schema, config);

  const result = oldVisit(ast, {
    leave: visitor,
  });

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const generated = result.definitions.filter((def) => typeof def === 'string');

  return {
    prepend: [`import * as yup from 'yup';`],
    content: generated.join('\n'),
  };
};
