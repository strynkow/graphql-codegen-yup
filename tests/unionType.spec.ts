import { buildSchema } from 'graphql';
import { plugin } from '../src/index';

describe('Nested types', () => {
  it('compiles inputs', async () => {
    const schema = buildSchema(/* GraphQL */ `
      type BlueEyesWhiteDragon0 {
        str: String!
      }
      type BlueEyesWhiteDragon1 {
        num: Int!
      }
      type BlueEyesWhiteDragon2 {
        num: Float!
      }
      # I play polymerization.
      union BlueEyesUltimateDragon = BlueEyesWhiteDragon0 | BlueEyesWhiteDragon1 | BlueEyesWhiteDragon2
    `);
    const result = await plugin(schema, [], {}, {});
    expect(result.content).toMatchSnapshot();
  });

  it('compiles inputs', async () => {
    const schema = buildSchema(/* GraphQL */ `
      type BlueEyesWhiteDragon0 {
        str: String!
      }
      type BlueEyesWhiteDragon1 {
        num: Int!
      }
      type BlueEyesWhiteDragon2 {
        num: Float!
      }
      # I play polymerization.
      union BlueEyesUltimateDragon = BlueEyesWhiteDragon0 | BlueEyesWhiteDragon1 | BlueEyesWhiteDragon2

      type Board {
        slot0: BlueEyesUltimateDragon!
        slot1: BlueEyesWhiteDragon1!
        slot2: BlueEyesWhiteDragon0!
      }
    `);
    const result = await plugin(schema, [], {}, {});
    expect(result.content).toMatchSnapshot();
  });
});
