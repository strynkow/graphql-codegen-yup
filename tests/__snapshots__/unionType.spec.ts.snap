// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested types compiles inputs 1`] = `
"export function BlueEyesWhiteDragon0Schema(): yup.SchemaOf<BlueEyesWhiteDragon0> {
  return yup.object({
    __typename: yup.mixed().oneOf(['BlueEyesWhiteDragon0', undefined]),
    str: yup.string().defined()
  })
}

export function BlueEyesWhiteDragon1Schema(): yup.SchemaOf<BlueEyesWhiteDragon1> {
  return yup.object({
    __typename: yup.mixed().oneOf(['BlueEyesWhiteDragon1', undefined]),
    num: yup.number().defined()
  })
}

export function BlueEyesWhiteDragon2Schema(): yup.SchemaOf<BlueEyesWhiteDragon2> {
  return yup.object({
    __typename: yup.mixed().oneOf(['BlueEyesWhiteDragon2', undefined]),
    num: yup.number().defined()
  })
}

export function BlueEyesUltimateDragonSchema(): any {
  return yup.lazy((value) => {
    switch (value?.__typename) {
      case 'BlueEyesWhiteDragon0':
        return BlueEyesWhiteDragon0Schema();
      case 'BlueEyesWhiteDragon1':
        return BlueEyesWhiteDragon1Schema();
      case 'BlueEyesWhiteDragon2':
        return BlueEyesWhiteDragon2Schema();
      default:
        throw new Error('Type not found.');
      }
  })
}
"
`;

exports[`Nested types compiles inputs 2`] = `
"export function BlueEyesWhiteDragon0Schema(): yup.SchemaOf<BlueEyesWhiteDragon0> {
  return yup.object({
    __typename: yup.mixed().oneOf(['BlueEyesWhiteDragon0', undefined]),
    str: yup.string().defined()
  })
}

export function BlueEyesWhiteDragon1Schema(): yup.SchemaOf<BlueEyesWhiteDragon1> {
  return yup.object({
    __typename: yup.mixed().oneOf(['BlueEyesWhiteDragon1', undefined]),
    num: yup.number().defined()
  })
}

export function BlueEyesWhiteDragon2Schema(): yup.SchemaOf<BlueEyesWhiteDragon2> {
  return yup.object({
    __typename: yup.mixed().oneOf(['BlueEyesWhiteDragon2', undefined]),
    num: yup.number().defined()
  })
}

export function BlueEyesUltimateDragonSchema(): any {
  return yup.lazy((value) => {
    switch (value?.__typename) {
      case 'BlueEyesWhiteDragon0':
        return BlueEyesWhiteDragon0Schema();
      case 'BlueEyesWhiteDragon1':
        return BlueEyesWhiteDragon1Schema();
      case 'BlueEyesWhiteDragon2':
        return BlueEyesWhiteDragon2Schema();
      default:
        throw new Error('Type not found.');
      }
  })
}

export function BoardSchema(): yup.SchemaOf<Board> {
  return yup.object({
    __typename: yup.mixed().oneOf(['Board', undefined]),
    slot0: BlueEyesUltimateDragonSchema().defined(),
    slot1: BlueEyesWhiteDragon1Schema().defined(),
    slot2: BlueEyesWhiteDragon0Schema().defined()
  })
}
"
`;
